### PR TITLE
Polish external reader API/implementation

### PR DESCRIPTION
**Pull request by translatenix**
Things I noticed but didn't address in this PR:
* many APIs are missing Javadoc
* libraries shouldn't log directly to std out/err (see `ExternalReaderProcessImpl`)
* some `ExternalReaderProcessException` messages are externalized, others aren't
* `ExternalReaderProcessException` is a checked exception
  - it's not clear to me what Pkl's policy on unchecked vs. checked exceptions is
* `new Random().nextLong()` is expensive and has poor randomness (Random should be reused or replaced with counter)
* deprecation of `ModuleKeyFactories.closeQuietly` in favor of `org.pkl.core.util.Readers.closeQuietly` turns package `org.pkl.core.util` into a public API
  - private packages keep leaking into public APIs
* `org.pkl.core.module.ResourceReaders` API is (too) tightly coupled to reader protocol via `ModuleReaderSpec`
* `ExternalReaderProcess` is a leaky abstraction
  - only partially implements the reader protocol
  - exposes its `MessageTransport`
  - interface name implies that all implementations spawn a process (unnecessary requirement?)
* msgpack is now a hard dependency of pkl-core, pkl-config-java, etc.
  - might be safer to make it an optional dependency
  
IMHO the external reader API/implementation needs more work.

Changes:
- keep implementation classes internal to their packages
- make classes final if possible
- make namespace classes non-instantiable
- throw IllegalStateException instead of ExternalReaderProcessException for use after close
  - common convention already used by HttpClient etc.
  - programming errors should be signaled with unchecked exceptions
- use private instead of public lock object
- polish Javadoc
- delete commented out code
- don't use star import for importing a single class

### CODE REVIEW

**Review by HT154 — APPROVED** [review]
Thanks! One little nit and one question from a non-native Java speaker.

File: pkl-core/src/main/java/org/pkl/core/externalreader/ExternalReaderProcessImpl.java
Comment (translatenix) [reply pkl-core/src/main/java/org/pkl/core/externalreader/ExternalReaderProcessImpl.java:53]: Encapsulating the locking protocol. Otherwise, users of the class (via the interface) might cause deadlocks by themselves synchronizing on `this`. You can find a better explanation in Effective Java 3rd Ed, Item 82.

File: pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/CliCommand.kt
Comment (translatenix) [reply pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/CliCommand.kt]: fixed

File: pkl-core/src/main/java/org/pkl/core/externalreader/ExternalReaderProcessImpl.java
Comment (HT154) [reply pkl-core/src/main/java/org/pkl/core/externalreader/ExternalReaderProcessImpl.java:53]: Gotcha, thanks!

**Comment by HT154** [comment]
A few clarifying questions about your unaddressed observations (inline)

> * many APIs are missing Javadoc

Aside from what you've addresses in this PR, which specifically? I'm thinking the added methods in `ModuleKeyFactories` and `ResourceReaders` at least, are there others?

> * libraries shouldn't log directly to std out/err (see `ExternalReaderProcessImpl`)

This mirrors pkl-server behavior today, but I agree that it's not ideal. Probably easy enough to address later on?

> * some `ExternalReaderProcessException` messages are externalized, others aren't

This seems easy enough to address, will open a follow-up PR.

> * `ExternalReaderProcessException` is a checked exception
>   
>   * it's not clear to me what Pkl's policy on unchecked vs. checked exceptions is

I'd love some core team input here as this is not an area I have much expertise in

> * `new Random().nextLong()` is expensive and has poor randomness (Random should be reused or replaced with counter)

Will address this too.

> * deprecation of `ModuleKeyFactories.closeQuietly` in favor of `org.pkl.core.util.Readers.closeQuietly` turns package `org.pkl.core.util` into a public API
>   
>   * private packages keep leaking into public APIs

Some core team input here as well would be great.

> * `org.pkl.core.module.ResourceReaders` API is (too) tightly coupled to reader protocol via `ModuleReaderSpec`

Not sure what you mean here, can you clarify? (typo?)

> * `ExternalReaderProcess` is a leaky abstraction
>   
>   * only partially implements the reader protocol

Same here, can you say more?

>   * exposes its `MessageTransport`

Ah yeah I agree and I'd really like to fix this before 0.27 releases. Having it vend the resolver instances directly seems like the best course of action.

>   * interface name implies that all implementations spawn a process (unnecessary requirement?)

Agreed, but a little wary of renaming the interface entirely.

> * msgpack is now a hard dependency of pkl-core, pkl-config-java, etc.
>   
>   * might be safer to make it an optional dependency

I'd prefer to defer to the core team on this.

**Comment by bioball** [comment]
Re: `Readers.closeQuietly`: yeah, we should definitely move it out of `org.pkl.core.util`. Probably fine to just put it into `org.pkl.core`, although I wonder if there's another place to put that method (`Readers` is a one-off class that we're not likely to add to)

---

Re: checked exceptions: generally, we throw checked exceptions for things that are internal to the implementation, that we then want to catch later and surface as `PklException` in our public APIs. We're not very consistent about applying this policy, though.

--- 

Re: msgpack: @translatenix: what are the concerns here? Is it the lack of responsiveness from the msgpack authors?

I think we will eventually want to move `BinaryEvaluator` into the core evaluator, which means we'll need a msgpack library one way or another. Curious about the MiniPack library that you mentioned earlier.

**Review by bioball — APPROVED** [review]
LGTM, good cleanups, thank you!

**Comment by translatenix** [comment]
@HT154 
> Aside from what you've addresses in this PR, which specifically? I'm thinking the added methods in ModuleKeyFactories and ResourceReaders at least, are there others?

The messaging package is largely undocumented. Is this now a public API?
(Would help tremendously if every package documented whether or not it's public.
Until Pkl adopts JPMS, this could  be done in package-info.java.)

> libraries shouldn't log directly to std out/err (see ExternalReaderProcessImpl)
>> This mirrors pkl-server behavior today, but I agree that it's not ideal. Probably easy enough to address later on?

Not my call. Personally I wouldn't ship a library that logs directly to std out/err.

> org.pkl.core.module.ResourceReaders API is (too) tightly coupled to reader protocol via ModuleReaderSpec
>> Not sure what you mean here, can you clarify? (typo?)

`org.pkl.core.resource.ResourceReaders.externalResolver()` accepts a `org.pkl.core.messaging.ResourceReaderSpec`, which models a binary message.
Hence the `org.pkl.core.resource` package now exports the `org.pkl.core.messaging` package, and changing a binary message in `org.pkl.core.messaging` may constitute a breaking API change in `org.pkl.core.resource`.

> ExternalReaderProcess is a leaky abstraction
>  * only partially implements the reader protocol
>> Same here, can you say more?

`ExternalReaderProcess` implements two messages and leaves the implementation of the remaining protocol to clients. Can't it implement the entire protocol instead of exposing its transport?

**Comment by translatenix** [comment]
@bioball 
> Re: msgpack: @translatenix: what are the concerns here? Is it the lack of responsiveness from the msgpack authors?

I'd be worried that users of pkl-config-java etc. may not want a 500 KB msgpack dependency that can also cause version conflicts in their builds. Especially because most users of pkl-config-java etc. won't actually need msgpack.

**Comment by bioball** [comment]
Yeah, fair. I'll put some more thought into vendoring that library, or making it an optional dependency.

> Not my call. Personally I wouldn't ship a library that logs directly to std out/err.

FYI: the writing to stderr there is an internal developer option; it's not documented and not something library users should ever expect to happen. It's a bit ugly that it exists, but it really helps with quality of life for maintainers.

**Pull request merged**
